File: Blind 75/07 - Trees/124-binary-tree-maximum-path-sum.py
'''
A path in a binary tree is a sequence of nodes where each pair of adjacent nodes in the sequence has an edge connecting them. A node can only appear in the sequence at most once. Note that the path does not need to pass through the root.

The path sum of a path is the sum of the node's values in the path.

Given the root of a binary tree, return the maximum path sum of any non-empty path.
'''

from math import inf
from time import time
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        res = [root.val]

        def dfs(node):
            if not node:
                return 0

            path_l = dfs(node.left)
            path_r = dfs(node.right)
            # Max path could be current node with left AND right branch
            res[0] = max(res[0], node.val + path_l + path_r)
            # Recursing up, path can only include left OR right path
            return max(node.val + max(path_l, path_r), 0)

        dfs(root)
        return res[0]

    def reference(self, root: Optional[TreeNode]) -> int:
        res = [root.val]

        # Return max path sum without split
        def dfs(root):
            if not root:
                return 0

            leftMax = dfs(root.left)
            rightMax = dfs(root.right)
            leftMax = max(leftMax, 0)
            rightMax = max(rightMax, 0)

            # Compute max path sum WITH split
            res[0] = max(res[0], root.val + leftMax + rightMax)
            return root.val + max(leftMax, rightMax)

        dfs(root)
        return res[0]

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.maxPathSum(case))
                else:
                    self.maxPathSum(case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        TreeNode(1, TreeNode(2), TreeNode(3)),
        TreeNode(-10, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7))),
    ]
    test.quantify(test_cases)

File: Blind 75/07 - Trees/JavaScript/124-binary-tree-maximum-path-sum.js
const { performance } = require("perf_hooks");

class TreeNode {
  constructor(val, left, right) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

class Solution {
  method(root) {
    const dfs = (node, path) => {
      if (!node) return 0;

      const leftGain = Math.max(dfs(node.left, path), 0);
      const rightGain = Math.max(dfs(node.right, path), 0);
      const currentMax = node.val + leftGain + rightGain;

      path[0] = Math.max(path[0], currentMax);

      return node.val + Math.max(leftGain, rightGain);
    };

    let maxPath = [-Infinity];
    dfs(root, maxPath);

    return maxPath[0];
  }

  reference(root, maxValue = [-Infinity]) {
    this.pathSum(root, maxValue);

    return maxValue[0];
  }

  pathSum = (root, maxValue) => {
    const isBaseCase = root === null;
    if (isBaseCase) return 0;

    return this.dfs(root, maxValue);
  };

  dfs = (node, maxValue) => {
    const left = Math.max(0, this.pathSum(node.left, maxValue));
    const right = Math.max(0, this.pathSum(node.right, maxValue));
    const sum = left + right + node.val;

    maxValue[0] = Math.max(maxValue[0], sum);

    return Math.max(left, right) + node.val;
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  new TreeNode(1, new TreeNode(2), new TreeNode(3)),
  new TreeNode(
    -10,
    new TreeNode(9),
    new TreeNode(20, new TreeNode(15), new TreeNode(7))
  ),
];
test.quantify(testCases);
